### PR TITLE
Feature: Linux Sockstat Plugin

### PR DESCRIPTION
**Pull request by paulkermann**
I picked up on the great work submitted in #597 (**linux socket stats**), applied the corrections proposed by the comments and had some refactor work in.
As the original PR has not seen action for several months, and would incur conflicts during merge, I created this one.

---

**Key changes:**
- The output now includes source/destination ports as columns (to avoid over-formatting).
- Missing values (or columns not supported by specific family) are returned as `NotApplicableValue`.
- The output is per task per socket - the aggregation was quite limiting in processing the output further, and doesn't greatly reduce the output volume.
- The supported socket family types are the same as the original since they satisfy my needs.

Please let me know if there are suggestions for improvement.

---

**Output sample:**
```
user@computer$ python3 ./vol.py -f /path/to/my/virtual/machine.vmem -r pretty linux.sockstat | grep INET
  |         Offset |      NetNS |   Pid |     Family |      Type | Proto                  |                                                                           Src Addr |   Src Port |                                                                           Dst Addr | Dst Port |       State | Extended
* | 0x90f07a958900 | 4026531992 |   960 |    AF_INET |     DGRAM |                    UDP |                                                                         127.0.0.53 |         53 |                                                                            0.0.0.0 |          | UNCONNECTED |         
* | 0x90f06bf1f1c0 | 4026531992 |   960 |    AF_INET |    STREAM |                    TCP |                                                                         127.0.0.53 |         53 |                                                                            0.0.0.0 |          |      LISTEN |         
* | 0x90f06c670980 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                                                 :: |      16443 |                                                                                 :: |          |      LISTEN |         
* | 0x90f02f680000 | 4026531992 |  5208 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      38044 |                                                                          127.0.0.1 |    16443 | ESTABLISHED |         
* | 0x90f06d225580 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                                   ::ffff:127.0.0.1 |      16443 |                                                                   ::ffff:127.0.0.1 |    38044 | ESTABLISHED |         
* | 0x90f079605f00 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                                                ::1 |      37140 |                                                                                ::1 |    16443 | ESTABLISHED |         
* | 0x90f078ba8980 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                                                 :: |      10250 |                                                                                 :: |          |      LISTEN |         
* | 0x90f02f6808c0 | 4026531992 |  5208 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      10248 |                                                                            0.0.0.0 |          |      LISTEN |         
* | 0x90f078b371c0 | 4026531992 |  5208 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      52856 |                                                                          127.0.0.1 |    16443 | ESTABLISHED |         
* | 0x90f06d224280 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                                                ::1 |      16443 |                                                                                ::1 |    37140 | ESTABLISHED |         
* | 0x90f06c675580 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                                                 :: |      10259 |                                                                                 :: |          |      LISTEN |         
* | 0x90f06c672f80 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                                                 :: |      10257 |                                                                                 :: |          |      LISTEN |         
* | 0x90f078b33d40 | 4026531992 |  5208 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      10256 |                                                                            0.0.0.0 |          |      LISTEN |         
* | 0x90f078b32300 | 4026531992 |  5208 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      10249 |                                                                            0.0.0.0 |          |      LISTEN |         
* | 0x90f06bf79c80 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                                   ::ffff:127.0.0.1 |      16443 |                                                                   ::ffff:127.0.0.1 |    52856 | ESTABLISHED |         
* | 0x90f06bf78980 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                              ::ffff:192.168.85.133 |      16443 |                                                              ::ffff:192.168.85.133 |    49148 | ESTABLISHED |         
* | 0x90f06bf7c280 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                              ::ffff:192.168.85.133 |      16443 |                                                              ::ffff:192.168.85.133 |    22281 | ESTABLISHED |         
* | 0x90f06bf7df00 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                              ::ffff:192.168.85.133 |      16443 |                                                              ::ffff:192.168.85.133 |    36888 | ESTABLISHED |         
* | 0x90f079656880 | 4026531992 |  5208 |   AF_INET6 |    STREAM |                    TCP |                                                              ::ffff:192.168.85.133 |      16443 |                                                                   ::ffff:10.1.90.1 |    45406 | ESTABLISHED |         
* | 0x90f06bf1d780 | 4026531992 |  5232 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      19001 |                                                                            0.0.0.0 |          |      LISTEN |         
* | 0x90f07255bd40 | 4026531992 |  5232 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      53406 |                                                                          127.0.0.1 |    19001 | ESTABLISHED |         
* | 0x90f0751f2bc0 | 4026531992 |  5232 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      19001 |                                                                          127.0.0.1 |    53406 | ESTABLISHED |         
* | 0x90f0777a5780 | 4026531992 |  5232 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      53408 |                                                                          127.0.0.1 |    19001 | ESTABLISHED |         
* | 0x90f06bf1cec0 | 4026531992 |  5232 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      19001 |                                                                          127.0.0.1 |    53408 | ESTABLISHED |         
* | 0x90f0751f3d40 | 4026531992 |  5244 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |       1338 |                                                                            0.0.0.0 |          |      LISTEN |         
* | 0x90f0513c5780 | 4026531992 |  5244 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      46435 |                                                                            0.0.0.0 |          |      LISTEN |         
* | 0x90f03212cc00 | 4026531992 |  5709 |   AF_INET6 |    STREAM |                    TCP |                                                                                 :: |      25000 |                                                                                 :: |          |      LISTEN |         
* | 0x90f00ed71a40 | 4026531992 |  7917 |    AF_INET |    STREAM |                    TCP |                                                                     192.168.85.133 |      53354 |                                                                       10.152.183.1 |      443 | ESTABLISHED |         
* | 0x90f00ed76040 | 4026531992 |  7918 |    AF_INET |    STREAM |                    TCP |                                                                     192.168.85.133 |      53358 |                                                                       10.152.183.1 |      443 | ESTABLISHED |         
* | 0x90f00ed71180 | 4026531992 |  7920 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |       9099 |                                                                            0.0.0.0 |          |      LISTEN |         
* | 0x90f00ed72bc0 | 4026531992 |  7920 |    AF_INET |    STREAM |                    TCP |                                                                     192.168.85.133 |      53356 |                                                                       10.152.183.1 |      443 | ESTABLISHED |         
* | 0x90f02abbdf00 | 4026531992 |  8176 |   AF_INET6 |    STREAM |                    TCP |                                                                                 :: |       9090 |                                                                                 :: |          |      LISTEN |         
* | 0x90f0593e71c0 | 4026531992 |  8176 |    AF_INET |    STREAM |                    TCP |                                                                          127.0.0.1 |      49278 |                                                                          127.0.0.1 |     9090 | ESTABLISHED |         
* | 0x90f06bf7f200 | 4026531992 |  8176 |   AF_INET6 |    STREAM |                    TCP |                                                                   ::ffff:127.0.0.1 |       9090 |                                                                   ::ffff:127.0.0.1 |    49278 | ESTABLISHED |         
* | 0x90f0751f0000 | 4026531992 |  8293 |    AF_INET |    STREAM |                    TCP |                                                                            0.0.0.0 |         22 |                                                                            0.0.0.0 |          |      LISTEN |         
* | 0x90f06bf79300 | 4026531992 |  8293 |   AF_INET6 |    STREAM |                    TCP |                                                                                 :: |         22 |                                                                                 :: |          |      LISTEN |         
* | 0x90f06bf18000 | 4026532664 |  8897 |    AF_INET |    STREAM |                    TCP |                                                                          10.1.90.1 |      45406 |                                                                       10.152.183.1 |      443 | ESTABLISHED |         
* | 0x90f079650980 | 4026532664 |  8897 |   AF_INET6 |    STREAM |                    TCP |                                                                                 :: |       9094 |                                                                                 :: |          |      LISTEN |    
```

### CODE REVIEW

File: volatility3/framework/plugins/linux/sockstat.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+                net_devices[netns_no][net_dev.ifindex] = dev_name
+
+        return net_devices
+
+    def get_unified_socket(self, sock, netns_no):
+        """
+        Takes a kernel generic `sock` object, and extracts the generic parameters - using the respective socket family.
+
+        Returns a tuple containing:
+            sock:           volatility object of the proper socket type (per socker family).
```
Comment (digitalisx) [inline volatility3/framework/plugins/linux/sockstat.py:78]: Typo
```suggestion
            sock:           volatility object of the proper socket type (per socket family).
```

File: volatility3/framework/plugins/linux/sockstat.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
+# This file is Copyright 2021 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+
+import logging
+from collections import defaultdict
+from typing import Callable, Tuple, List, Dict
```
Comment (digitalisx) [inline volatility3/framework/plugins/linux/sockstat.py:7]: Unused module
```suggestion
from typing import Callable, Dict, List
```

File: volatility3/framework/plugins/linux/sockstat.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+from volatility3.framework.renderers import NotApplicableValue, format_hints
+from volatility3.framework import renderers, interfaces, exceptions, constants, objects
```
Comment (digitalisx) [inline volatility3/framework/plugins/linux/sockstat.py:10]: Unused module
```suggestion
from volatility3.framework import constants, exceptions, interfaces, renderers
```

**Review by digitalisx — COMMENTED** [review]
Hello @paulkermann,
I left some NIT comments :)

**Comment by eve-mem** [comment]
Hey @paulkermann is this pull request still needed with the sockstat plugin being merged in a different pull request?

**Comment by ikelos** [comment]
Closing due to inactivity, feel free to reopen it if you feel the PR is still relevant.

**Pull request closed without merging**